docs/spec/overload.rst

```diff
@@ -99,8 +109,349 @@ variable is not sufficient.
+be expanded into ``int`` and ``str``.
+
+Type expansion should be performed one argument at a time from left to
+right. Each expansion results in sets of effective argument types.
+For example, if there are two arguments whose types evaluate to
+``int | str`` and ``int | bytes``, expanding the first argument type
+results in two sets of argument types: ``(int, int | bytes)`` and
+``(str, int | bytes)``. If type expansion for the second argument is required,
+four sets of argument types are produced: ``(int, int)``, ``(int, bytes)``,
+``(str, int)``, and ``(str, bytes)``.
+
+After each argument's expansion, return to step 2 and evaluate all
+expanded argument lists.
+
+- If all argument lists evaluate successfully, combine their
+  respective return types by union to determine the final return type
+  for the call, and stop.
+- If argument expansion has been applied to all arguments and one or
+  more of the expanded argument lists cannot be evaluated successfully,
+  generate an error and stop.
+
+
+For additional details about argument type expansion, see
+`argument-type-expansion`_ below.
+
+
+Step 4: If the argument list is compatible with two or more overloads,
+determine whether one or more of the overloads has a variadic parameter
+(either ``*args`` or ``**kwargs``) that maps to a corresponding argument
+that supplies an indeterminate number of positional or keyword arguments.
+If so, eliminate overloads that do not have a variadic parameter.
+
+- If this results in only one remaining candidate overload, it is
+  the winning match. Evaluate it as if it were a non-overloaded function
+  call and stop.
+- If two or more candidate overloads remain, proceed to step 5.
+
+
+Step 5: For each argument, determine whether all possible
+:term:`materializations <materialize>` of the argument's type are assignable to
+the corresponding parameter type for each of the remaining overloads. If so,
+eliminate all of the subsequent remaining overloads.
+
+For example, if the argument type is ``list[Any]`` and there are three remaining
+overloads with corresponding parameter types of ``list[int]``, ``list[Any]``
+and ``Any``. We can eliminate the third of the remaining overloads because
+all materializations of ``list[Any]`` are assignable to ``list[Any]``, the
+parameter in the second overload. We cannot eliminate the second overload
+because there are possible materializations of ``list[Any]`` (for example,
+``list[str]``) that are not assignable to ``list[int]``.
+
+Once this filtering process is applied for all arguments, examine the return
+types of the remaining overloads. If these return types include type variables,
+they should be replaced with their solved types. If the resulting return types
+for all remaining overloads are :term:<equivalent>, proceed to step 6.
+
+If the return types are not equivalent, overload matching is ambiguous. In
+this case, assume a return type of ``Any`` and stop.
```

Comment:
Is there a correctness requirement for the return type here to be assumed to be `Any`? It seems to me that it would also be valid for a type-checker to use the union of all the ambiguous matching overloads. I would prefer for the specification not to prevent that option.

Reply:
Existing stubs (including typeshed, numpy, pandas, and others) assume that the result will be `Any` in this case, so I don't think this is something we can change at this point. An earlier version of pyright generated a union, and it resulted in many false positive errors and lots of unhappy users. I think it's important for the spec to specify `Any` here so stub authors can rely on the behavior.

Reply:
Hmm, ok, that's useful info, thank you. If you happen to know of any old pyright issues where unhappy users surfaced these problems with the union behavior, I would be curious to take a look at some real world cases relying on this.

Reply:
I am not convinced that this is a point where it is in fact important for "stub authors to be able to rely on the behavior." The responsibility of stub authors is to accurately model their overloads. If a union (instead of `Any`) return type causes a false positive at a call site, and the overloads are correct, that's not a concern of the stub author, it's an issue of imprecise (dynamic) types at the call site, where perhaps some invariants are known but not expressed in an argument type. I think this is a reasonable candidate for type checkers to offer different levels of strictness depending on user preference.

If you have examples where a stub author would write their overloads differently because they are depending on this behavior, I would like to see those examples.

Reply:
> I guess the precise construct one would need here in order to return a more precise type is something that I think has been requested before under the name `AnyOf`; effectively a "gradual union", or a gradual type which can materialize to any element of the union. (I think this type could also be formulated, given intersections, as the union of each element intersected with `Any`.)

That's a very interesting idea. If you consider `Any` as an opportunistic superposition of all possible types, then this gradual union would be an opportunistic superposition of the types within it. So in that sense it's a very natural way to generalize `Any`, by restricting it to only be able to assume a given subset of types. Nice!

Reply:
Interestingly, pyright internally generates an `AnyOf` type in this circumstance. From a typing perspective, it acts like an `Any`, but the additional information is used behind the scenes to provide better completion suggestions in the language server. For example, in this case the revealed type is `Any`, but the completion suggestions reflect the fact that the value must be of type `AnyOf[str, int]`.

<img width="360" alt="image" src="https://github.com/user-attachments/assets/625374aa-a053-431d-a02a-9f52d5ea489c" />